tools/rewriter/GenCallAsm.cpp

```diff
@@ -652,19 +952,22 @@ std::string emit_asm_wrapper(const CAbiSignature &sig,
     // precedes the stack return itself.
     start_of_ret_space += 8;
 
-    // See what unpadded alignment would be for the start of the stack return.
-    size_t unaligned = start_of_ret_space % stack_return_align;
+    // Align the size of the return space to `stack_return_align`.
+    size_t unaligned = (start_of_ret_space + stack_return_size) % stack_return_align;
     stack_return_padding = unaligned != 0 ? stack_return_align - unaligned : 0;
   }
 
   // Count room for for the ret align padding, return value, and our ret ptr.
   size_t compartment_stack_space = start_of_ret_space + stack_return_size +
-                                   stack_return_padding + stack_arg_size;
-  // Compute what the stack alignment would be before calling the wrapped
-  // function to check if we need to insert 8 bytes for alignment. We add 8
-  // bytes to the compartment_stack_space since the frame is initially off by 8
-  // bytes.
-  size_t stack_alignment = (compartment_stack_space + 8) % 16;
+                                   stack_return_padding + stack_arg_size + stack_arg_padding;
+  size_t stack_alignment = 0;
+  if (arch == Arch::X86) {
```

Comment:
I think we have the alignment block that this corresponds to on both the arm and x86 diagrams so I don't see why this check is x86-specific.

Reply:
We don't actually insert it on ARM. I'll try to figure out some of the TODOs on the diagram.